### PR TITLE
Adds ^board command to pyinfo cog

### LINKED ISSUE #1769: [Reserved for Alianin] Create links to dawnglare boards with ^board
* `^board GGGGGG GGGGGG GGGGGG GGGGGG GGGGGG` should link you to https://pad.dawnglare.com/?patt=GGGGGGGGGGGGGGGGGGGGGGGGGGGGGG&showfill=1
  * Fill direction is RIGHT then DOWN
  * Validate that there are either 20 (5x4) or 30 (6x5) or 42 (7x6) letters specified
  * If it's 5x4 or 7x6 then add the appropriate board size to the URL (&height=6&width=7 or &height=4&width=5)
  * Remove all spaces
  * Also include the param `&showfill=1` at the end of the url
* `^board -1 LLLLL DDDDD LLLLL DDDDD LLLLL DDDDD` should link you to https://pad.dawnglare.com/?patt=LDLDLDLDLDLDLDLDLDLDLDLDLDLDLD&showfill=1 - the -1 at the start causes it to first invert the pattern and then fill it
  * The spaces are irrelevant to the inversion, it's just the `-1` causing this
  * I think you should do this with matrices but I'm not sure?
  * Fill direction is DOWN then RIGHT
  * Same thing about board sizes

Allowed letters: `RGBLDHXJMPZ`, if any other letter is included, use `send_cancellation_message` (this comes from the library `tsutils` which is on PyPI) to inform that the user has used an invalid color

This should go in the `padinfo` cog in padinfo.py!

### PR DESCRIPTION
**Pull request by Alianin**
Generates a Dawnglare link from a user provided string with optional
fill inversion.

Resolves #1769

### CODE REVIEW

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:18]: This whole function can be super well condensed into the one line of
```py
return BoardGenerator.inversion in self.board
```

And at that point, we don't even need to keep it as a function! It's so short and clear that we can just put it up into the \_\_init\_\_ :)

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:35]: Even if we're going to define it in a function that \_\_init\_\_ calls, we should always define all of our variables in the \_\_init\_\_ with some default value.  If we know we're definitely going to set it in a subfunction, we can set it to None initially, but it's always good to know exactly what attributes a class has just by looking in the  \_\_init\_\_!

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:26]: We generally want to restrict our function names to use a set of a few verbs.  Some that I like are `get`, `set`, and `validate` which I think are super clear!

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:42]: I would put my local substring right here rather than at the top!  I would put `build_string = ""` and then `build_string += board[j][i]` in the loop right after!  That means there's no need for a second `clean_string` variable because our `build_string` is already a string. :)

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:53]: Same as my comment for `inversion_check`!  This too could be condensed into a single line that uses the fact that `all` already returns a boolean value.

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:81]: This works, but with a function like this as well as a few others, I'd have it return the `link variable` and set `self.link` to it in `__init__` rather than down here.

File: padinfo/boardgenerator.py
Comment (chasehult) [inline padinfo/boardgenerator.py:65]: Fancy Python syntax: you can set two variables at once with `height, width = 4, 5`!  Totally optional, but I think it's easier to read :)

**Review by chasehult — CHANGES_REQUESTED** [review]

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [inline padinfo/boardgenerator.py:23]: EXTREMELY OPTIONAL!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

okay so, what you have here is TOTALLY FINE and works. But, there's a small "separation of concerns" that I don't love, and you can either do it now, or, you could come back in a couple projects from now and redo it later. If it doesn't 100% make sense now I'd rather you do it when it does make sense instead of doing it just because I'm saying so.

* In the `init` you get `self.invert` as whether `-1` is in the string (btw, I LOVE that you set that as a class constant instead of hard-coding the string `-1` in the lookup!!). That on its own is fine, but
* Then here, in a totally different location, we're dealing with the consequences of that inversion being there. Now, this is the ONLY place where we're dealing with those consequences.
	* If there were multiple places where we cared about the fact that the inversion were present, it might make sense to do a dedicated class constant and use it in multiple places
	* Also we could potentially set a class constant for e.g. debugging purposes, so it may make sense to set the class constant to None in the init and then set the correct value here
	* The problem is, we've detected the presence of the -1 way far away in the init, and then we're dealing with it here, plus modifying the string here, and so these two bits of logic are really separate from each other. If we were to change how setting inversion were done (say in the future it was done by wrapping the thing in `()`) then the pieces of logic are super far away from each other
* So, I would like to see a refactor that brought those pieces of logic closer to each other. Again the class constant isn't necessarily a bad idea if you wanna debug, but I'd detect inversion presence near the rest of the inversion logic (and then set the value of the constant there)

Again there is absolutely no need to resolve this issue right now, and if you don't like 100% totally understand it, please postpone it until you do!!

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [inline padinfo/boardgenerator.py:60]: You don't have to set local variables before defining the class constants, it's ok to immediately:

`self.board_height, self.board_width = 4, 5`

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [inline padinfo/boardgenerator.py:48]: again this is pretty optional, but this could be done via a class constant with a dictionary if you want. `board_sizes = { }` and I'll let you see if you can figure out the rest :)

You'd want to use the dict.get() method which returns `None` on a missing input so that you can set `invalid_board` in that case instead of getting a `KeyError`

(this is optional!!! it can also be a later PR!)

File: padinfo/padinfo.py
Comment (RheingoldRiver) [inline padinfo/padinfo.py:1421]: let's uppercase the message before parsing it so that people can input lowercase letters and be lazy

File: padinfo/padinfo.py
Comment (RheingoldRiver) [inline padinfo/padinfo.py:1427]: If someone does an invalid board AND has invalid letters, you probably wanna return both errors. So this shouldn't be an elif here. Then you can send the link if there were no errors

File: padinfo/boardgenerator.py
Comment (Alianin) [reply padinfo/boardgenerator.py:60]: I was doing that to make the declarations shorter for functions where I was using that variable a lot. But if that's against best practices, I can change it.

Comment (RheingoldRiver) [reply padinfo/boardgenerator.py:60]: ah, I'd say that's superseded by "avoid unnecessary local variables." This is a bit of a judgment call. Elsewhere you aliased `board = self.board` and I didn't say "don't do that" because I know you did that to make it easier to read, and probably in 6 months you won't find it necessary to do that, but for now, seeing long "variable" names like `self.board` can actually make the code harder to read and `board = self.board` is arguably necessary.

If you have something like `data["player"]["stats"]["item"]` and you're referencing it 3 times in the same function, to get `data["player"]["stats"]["item"]["id"]` and `data["player"]["stats"]["item"][is_null]` and `data["player"]["stats"]["item"]["name"]` then I will 100% alias this as `item = data["player"]["stats"]["item"]` at the start because that's way too hard to read.

so this is a tradeoff between "how long is an annoyingly long parameter to read vs avoid local variable names"

Having only `self.param_name` vs `param_name` is generally considered *not* so long that you have to alias it, and especially in this case, the presence of your local variable names leads the reader of your code to think that you will be using the local variables rather than just assigning them at the end of the function and then exiting

File: padinfo/padinfo.py
Comment (RheingoldRiver) [inline padinfo/padinfo.py:1425]: Wait this isn't true, is it? You delete all of the spaces immediately right? (Sorry, this is probably my fault in the ticket - I put spaces as an example to make it clearer to understand but if someone wanted to just put a string without any spaces that should be valid input, and if they wanted to do like `GGG GGG GGG GGG GGG GGG...` that would also be fine. All spaces should get immediately stripped. But I'm pretty sure, you're already doing this.)

Comment (Alianin) [reply padinfo/padinfo.py:1425]: Having spaces makes it infinitely easier for me to handle the input. I'm stripping the spaces when I convert the input to a list. It doesn't matter for non-inverted, but for inverted it means I don't have to do fancy calculations to split them out into correctly numbered lists for the inverted fill.

It'll probably be easier to handle with some changes I'm making. I'll commit those now for review and if it matters that we need to accept non-spaced inputs I'll work on getting that sorted.

File: padinfo/boardgenerator.py
Comment (Alianin) [reply padinfo/boardgenerator.py]: I did this a little differently than you suggested but it seems clean with a lot less code.

File: padinfo/boardgenerator.py
Comment (Alianin) [reply padinfo/boardgenerator.py]: To be clear, do you mean this line in get_clean_board? `self.board.remove(self.inversion)`

I'm already thinking through some refactors based on your comments on being able to take an input not separated with spaces. If that line is the problem, then I think I know how to resolve this, but I want to be sure first.

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [reply padinfo/boardgenerator.py]: this is great!! the only thing is, in contrast to what I said about "avoiding local variables," I *personally* prefer to avoid dict calls inside of fstrings because having quotes inside of quotes kind of sucks (imo). Some people don't care about that and will do what you did. But I actually would say `height = self.board_sizes[self.board_length]['height']` and then sub `height` into the fstring now, and same thing for width. This is a bit personal-preference-y though.

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [reply padinfo/boardgenerator.py]: So the problem isn't any one thing in isolation, but that

```
self.invert = self.inversion in self.board```

is very far away from

```
        if not self.invert:
            return ''.join(self.board)

        self.board.remove(self.inversion)
```

AND ALSO

that `self.invert` is never used except in conjunction with this other block of code.

So separating the assignment of `self.invert` (and therefore the logic to determine its value) so very far away from where we're using it is kinda sus, because there's no real reason to.

If we were using `self.invert` all over the place, then by necessity these things might be far away from each other, but, that's not the case here, so we should do them close to each other.

File: padinfo/padinfo.py
Comment (RheingoldRiver) [reply padinfo/padinfo.py:1425]: Ahhh got it. See if you can figure out a way to split the string based on the board dimensions, but if not, then can you try and require spaces iff it's inverted?

File: padinfo/padinfo.py
Comment (RheingoldRiver) [inline padinfo/padinfo.py:1415]: sorry one small change that's also my fault - can you rename this function to `boardlink` ? Aradia reminded me that we have another command called `boards` and `board` autocorrects to it (you don't have the cog it's in loaded so it hasn't been a problem for you) whoops

You can leave this comment open & do it at the very end if you're used to typing ^board for testing so you don't have to change

File: padinfo/boardgenerator.py
Comment (Alianin) [reply padinfo/boardgenerator.py]: Take a look at my latest commit and see if it resolves your concern. I did a bit bigger refactor with the inversion logic to handle not having spaces in the input. I'm stripping the inversion characters right away after I check for its presence so that I can do the rest of the necessary calculations on a clean string.

File: padinfo/boardgenerator.py
Comment (Alianin) [reply padinfo/boardgenerator.py]: I reworked this one more time and took your advice on happy path and return ASAP.

File: padinfo/padinfo.py
Comment (Alianin) [reply padinfo/padinfo.py]: The latest commit should resolve this now.

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [reply padinfo/boardgenerator.py]: yep, this looks a lot better now

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [inline padinfo/boardgenerator.py:18]: I'd rename this to `invalid_size` since both invalid size & invalid letters are invalid boards so that's a more specific error name

File: padinfo/padinfo.py
Comment (RheingoldRiver) [inline padinfo/padinfo.py:74]: this should be an explicit path, `from padinfo/boardgenerator import BoardGenerator`

although actually you should rename `boardgenerator` to `board_generator`

two notes about this:

1. In Pycharm to rename a file safely, right-click on it and go to Refactor --> Rename and then type the new name, and it will fix all instances of you having imported from it automatically (although with the `.boardgenerator` syntax I'm not 100% sure if it will detect this one tbh but do it like that anyway)
2. Ok so to import a file, in your code you write like, `board = BoardGenerator(message.upper())` and then `BoardGenerator` has a squiggly red underline on it right. So then you press `alt+Enter` underneath `BoardGenerator` and itll be like `Import BoardGenerator....` and you're like YES! and then you pick the one you want to import (sometimes there's multiple options, e.g. if you wrote your own plus you have installed a library that generates one, etc). Then, it'll auto-add the library to the imports, with the full path written out, and you never had to scroll to the top of the file to do it. Super op!
    * So, to fix the relative path import here, delete the import altogether, and then press alt+Enter on BoardGenerator, and use this keyboard shortcut.

File: padinfo/boardgenerator.py
Comment (RheingoldRiver) [inline padinfo/boardgenerator.py:20]: you can do a happy-path return here too - if there ARE errors then return, otherwise de-indent and continue doing stuff

**Review by RheingoldRiver — COMMENTED** [review]
Okay I'm pretty sure these are the last 3 things and they're all really small at this point! Sorry I'm a day late on the review & also sorry for not catching them earlier. Your code is really good now!! I'll make Aradia check it over one more time before merging too, just to make sure I'm not missing anything but we should be good to go after this!

**Review by RheingoldRiver — APPROVED** [review]

**Review by chasehult — APPROVED** [review]

**Pull request merged**
